imap_processing/hit/l0/decom_hit.py

```diff
@@ -0,0 +1,463 @@
+
+    # Find indices where sequence flag is 1 (the start of a science frame)
+    # and filter for indices that are 20 packets apart. These will be the
+    # starting indices for science frames in the science data.
+    start_indices: np.array = np.where(seq_flgs == 1)[0]
+    valid_start_indices = start_indices[np.where(np.diff(start_indices) == 20)[0]]
+    last_index_of_frame = None
+
+    if valid_start_indices[0] != 0:
+        # The first start index is not at the beginning of the file.
+        print(
+            f"{valid_start_indices[0]} packets at start of file belong to science frame from previous day's ccsds file"
+        )
+        # TODO: Will need to handle these packets when processing multiple files
+
+    for i, start in enumerate(valid_start_indices):
+        # Get sequence flags and counters corresponding to this science frame
+        seq_flgs_chunk = seq_flgs[start : start + packets_in_frame]
+        src_seq_ctr_chunk = src_seq_ctrs[start : start + packets_in_frame]
+
+        # Check for valid science frames with proper sequence flags and counters
+        # and append corresponding science data to lists.
+        if is_valid_science_frame(seq_flgs_chunk, src_seq_ctr_chunk):
+            science_data_chunk = science_data[start : start + packets_in_frame]
+            epoch_data_chunk = epoch_data[start : start + packets_in_frame]
+            # First 6 packets contain count rates data
+            count_rates_binary.append("".join(science_data_chunk[:6]))
+            # Last 14 packets contain pulse height event data
+            pha_binary.append("".join(science_data_chunk[6:]))
+            # Just take first packet's epoch for the science frame
+            epoch_science_frame.append(epoch_data_chunk[0])
+            last_index_of_frame = start + packets_in_frame
+        else:
+            # TODO: log issue
+            # Skip invalid science frame and move on to the next one
+            print(
+                f"Invalid science frame found with starting packet index = " f"{start}"
+            )
+
+    if last_index_of_frame:
+        remaining_packets = total_packets - last_index_of_frame
+        if remaining_packets < packets_in_frame:
+            # TODO: log extra packets at end of file.
+            #  Need to handle these packets that belong to the next day's science frame.
+            print(
+                f"{remaining_packets} packets at end of file belong to science frame from next day's ccsds file"
+            )
+
+    # Add new data variables to the dataset
+    epoch_science_frame = np.array(epoch_science_frame)
+    sci_dataset = sci_dataset.drop_vars("epoch")
+    sci_dataset.coords["epoch"] = epoch_science_frame
+    sci_dataset["count_rates_binary"] = xr.DataArray(
+        count_rates_binary, dims=["epoch"], name="count_rates_binary"
+    )
+    sci_dataset["pha_binary"] = xr.DataArray(
+        pha_binary, dims=["epoch"], name="pha_binary"
+    )
+    return sci_dataset
```

Comment:
Now, you need to figure out how to filter rest of data variables in `sci_dataset` to filter out only good science frame's data. Otherwise you will have different data shape between `count_rates_binary`, `pha_binary`, and `epoch` vs CCSDS header variables. You can use `.isel` on `sci_dataset`.